### ISSUE TITLE
[bug] package-id doesn't match the locked one (full_transitive_package_id and header_only)

### ISSUE THREAD

**Issue by spoofedex** [conan-io/conan#8296]
Hi all,

I've run into an issue, of which I've managed to create a minimal reproduction. I'm not completely sure if this is a bug in Conan, or in my use case. It is triggered when a package calls `self.info.header_only()` in the `package_id` method of a conanfile, while full_transitive_package_id is set to True. I'm not sure if this combination even makes sense, but the former was a left-over from the past, and full_transitive_package_id was enabled at a later point.
For now, we will simply remove the call to header_only; I believe this to be a suitable workaround. However, I am still submitting this so that you can determine whether this is a bug or not. Also, it may save people some time!

Best regards,
SpoofedEx

(PS: thanks for the fast response on my last ticket(s)!)

### Environment Details (include every applicable attribute)
  * Operating System+version: Windows 10
  * Compiler+version: N.A.
  * Conan version: 1.32.1
  * Python version: 3.7.3

### Steps to reproduce (Include if Applicable)
Have the following settings:
```
revisions_enabled = True
default_package_id_mode = package_revision_mode
full_transitive_package_id = True
```

Consider the following packages:

pkga/conanfile.py:
```
from conans import ConanFile

class PkgaConan(ConanFile):
    name = "pkga"
    version = "1.0"
    requires = "pkgb/1.0@test/test", "pkgd/1.0@test/test"
```

pkgb/conanfile.py:
```
from conans import ConanFile

class PkgbConan(ConanFile):
    name = "pkgb"
    version = "1.0"
    requires = "pkgc/1.0@test/test"
```

pkgc/conanfile.py:
```
from conans import ConanFile

class PkgcConan(ConanFile):
    name = "pkgc"
    version = "1.0"
    requires = "pkgd/1.0@test/test"
```

pkgd/conanfile.py:
```
from conans import ConanFile

class PkgdConan(ConanFile):
    name = "pkgd"
    version = "1.0"
    requires = "pkge/1.0@test/test"

    def package_id(self):
        self.info.header_only()
```

pkge/conanfile.py:
```
from conans import ConanFile

class PkgeConan(ConanFile):
    name = "pkge"
    version = "1.0"
    options = { "test": [True, False] }
    default_options = { "test": True }
```

Execute the following commands:
```
conan create pkge test/test
conan create pkgd test/test
conan create pkgc test/test
conan create pkgb test/test
conan export pkga test/test
conan lock create --reference pkga/1.0@test/test --lockfile-out conan.lock --build missing --build outdated --build cascade
conan install pkga/1.0@test/test --build pkga --lockfile conan.lock
```

The last command will fail with the following output:
```
ERROR: 'pkgb/1.0@test/test#4e25de8f3298517d1a7aac485bf7bcfc' package-id '21de809762641568c6acfc78c41460b06c1d36e7' doesn't match the locked one '1cf745521f781fcf1bf8f138b2d3471202bf3269'
```

What I've seen in my diagnosis: during the creation of the lockfile, the requirements (not the "full requirements") are used for transitively propagating options. Since pkgd calls `self.info.header_only()`, pkge ends up in pkgd's full requirements, but not in its requirements. As such pkge's options will be propagated to pkgd, but not to pkgc and pkgb. For some reason (I haven't looked into this) it does get propagated to pkga (the root), which also depends on pkgd. These options are written to the lockfile:
```
{
 "graph_lock": {
  "nodes": {
   "1": {
    "ref": "pkga/1.0@test/test#8a31ef63b0abf183dc216424b2c0fba0",
    "options": "pkge:test=True",
    "package_id": "5c35c09b41f39ac11d0e2a452ddf594995429056",
    "requires": [
     "2",
     "4"
    ],
    "context": "host"
   },
   "2": {
    "ref": "pkgb/1.0@test/test#4e25de8f3298517d1a7aac485bf7bcfc",
    "options": "",
    "package_id": "1cf745521f781fcf1bf8f138b2d3471202bf3269",
    "prev": "33a5d729eda060dd9e8c6c4a5b2109ae",
    "requires": [
     "3"
    ],
    "context": "host"
   },
   "3": {
    "ref": "pkgc/1.0@test/test#4222a2863e0db5b9bcc2fd7b1c9f5272",
    "options": "",
    "package_id": "4ed253ec0d52b41a3aa53f3ffb6810c1eb81feb8",
    "prev": "c47970819e28765e7404fe67bf2e0ddd",
    "requires": [
     "4"
    ],
    "context": "host"
   },
   "4": {
    "ref": "pkgd/1.0@test/test#1bf270348b9a875c2f3b0f0ef6858ac1",
    "options": "pkge:test=True",
    "package_id": "5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9",
    "prev": "04593d002e88f8e523f01fd160081e86",
    "requires": [
     "5"
    ],
    "context": "host"
   },
   "5": {
    "ref": "pkge/1.0@test/test#c8b4e4f7ccfcf7d77995132efeb33fd0",
    "options": "test=True",
    "package_id": "3c106a5cf48458f46f490a1ac414b8c6446e70b8",
    "prev": "1aff2c5afe664ad771a25764a7d70298",
    "context": "host"
   }
  },
  "revisions_enabled": true
 },
 "version": "0.4",
 "profile_host": "[settings]\narch=x86_64\narch_build=x86_64\nbuild_type=Debug\ncompiler=Visual Studio\ncompiler.runtime=MDd\ncompiler.version=14\nos=Windows\nos_build=Windows\n[options]\n[build_requires]\n[env]\n"
}
```

As a result, pkgb does not have any options, but pkga does. When compiling pkga, the options will be propagated to pkgb. So the lockfile has a version of pkgb without options, whereas during compiling it does expect a version with options.

This causes an issue if (and only if) full_transitive_package_id is set to True, as (apparently) these options are then taken into account in the calculation of the package_id. This causes a mismatch in the package_id between the lockfile and what is expected at the time of compilation.

**Comment by jgsogo** [comment]
Hi, @spoofedex 

Thanks a lot for the detailed explanation and all the sources needed to reproduce it. Indeed I see different behavior when creating the lockfile (or when running `conan create/install`)  and when using it.

When creating the lockfile (or working without lockfiles), the option `pkge:test=True` is not considered in the computation of `pkgb` package-id, but when using the lockfile it is. I keep having a look at it 🕵️

**Comment by jgsogo** [comment]
I see it is related to the propagation of options from indirect requirements. I did some changes [here](https://github.com/conan-io/conan/compare/develop...jgsogo:issue/8296-transitive-reqs?expand=1) and it works, but it would break existing behavior. I did it just to understand how the propagation of this information works.

I'm sorry I don't have an answer right now for this issue, we need @memsharded here to know if this _propagation logic_ is a `bug`, a `known-bug`, a `feature` or we can approach it from a different perspective.

For sure this is a `bug` related to lockfiles, but I don't know how much we can break or not in order to fix it.

Thanks for your patience.

**Comment by memsharded** [comment]
I am trying to reproduce this, but so far without success:

```python
  def test_options_error(self):
        # https://github.com/conan-io/conan/issues/8296
        client = TestClient()
        client.run("config set general.revisions_enabled=True")
        client.run("config set general.default_package_id_mode=package_revision_mode")
        client.run("config set general.full_transitive_package_id=True")
        client.save({"pkga/conanfile.py": GenConanfile().with_requires("pkgb/1.0", "pkgd/1.0"),
                     "pkgb/conanfile.py": GenConanfile().with_requires("pkgc/1.0"),
                     "pkgc/conanfile.py": GenConanfile().with_requires("pkgd/1.0"),
                     "pkgd/conanfile.py": GenConanfile().with_package_id("self.info.header_only()"),
                     "pkge/conanfile.py": GenConanfile().with_option("test", [True, False])
                    .with_default_option("test", True)
                     })
        for pkg in "pkge", "pkgd", "pkgc", "pkgb":
            client.run("create {pkg} {pkg}/1.0@".format(pkg=pkg))
        client.run("export pkga pkga/1.0@")
        client.run(
            "lock create --reference pkga/1.0@ --lockfile-out conan.lock --build missing --build outdated --build cascade")
        print(client.load("conan.lock"))
        client.run("install pkga/1.0@ --build pkga --lockfile conan.lock")
```
This works fine, it doesn't error.
The lockfile does not contain the options part:
```
"graph_lock": {
  "nodes": {
   "1": {
    "ref": "pkga/1.0",
    "options": "",
    "package_id": "4e4db41bce1354918da6e4557b952f049559a0dd",
    "requires": [
     "2",
     "4"
    ],
    "context": "host"
   },
   "2": {
    "ref": "pkgb/1.0",
    "options": "",
    "package_id": "6368ae9b5c48314e4260fcb0d3d8de7e3e499860",
    "prev": "0",
    "requires": [
     "3"
    ],
    "context": "host"
   },
...
```
There might be some other detail that I am missing, could you please have a look and check? Thanks.

**Comment by jgsogo** [comment]
@memsharded , you are missing `pkgd -> pkge` requirement.

**Comment by memsharded** [comment]
Thanks, able to reproduce with:

```python
    def test_options_error(self):
        # https://github.com/conan-io/conan/issues/8296
        client = TestClient()
        client.run("config set general.full_transitive_package_id=True")
        client.save({"pkga/conanfile.py": GenConanfile().with_requires("pkgb/1.0", "pkgd/1.0"),
                     "pkgb/conanfile.py": GenConanfile().with_requires("pkgc/1.0"),
                     "pkgc/conanfile.py": GenConanfile().with_requires("pkgd/1.0"),
                     "pkgd/conanfile.py": GenConanfile().with_package_id("self.info.header_only()")
                                                        .with_requires("pkge/1.0"),
                     "pkge/conanfile.py": GenConanfile().with_option("test", [True, False])
                    .with_default_option("test", True)
                     })
        for pkg in "pkge", "pkgd", "pkgc", "pkgb":
            client.run("create {pkg} {pkg}/1.0@".format(pkg=pkg))
        client.run("export pkga pkga/1.0@")
        client.run("lock create --reference pkga/1.0@ --lockfile-out conan.lock --build missing")
        print(client.load("conan.lock"))
        client.run("install pkga/1.0@ --build pkga --lockfile conan.lock")
```

The package_id_mode and revisions do not affect, it is the full_transitive_package_id that is affecting (removing it solves the issue too)

**Comment by memsharded** [comment]
I have been investigating this and seems very challenging to fix. The changes done in https://github.com/conan-io/conan/compare/develop...jgsogo:issue/8296-transitive-reqs?expand=1 are too breaking, they will change the existing package_ids of many users. I am checking other alternatives, but seems a really bad alignment of full_transitive_package_id + options + lockfiles + header_only, not sure what will be possible.

**Comment by spoofedex** [comment]
Hi all,

Thanks for the quick and good responses, once again!
For us, we have simply worked around it by removing the header_only call. This may have have some small impact on performance/disk usage, but I don't expect it to cause any issues. So we're not in a rush for a solution.

Best regards,
SpoofedEx

**Comment by memsharded** [comment]
Solved in Conan 2.0, closing.
